File: array/560. Subarray Sum Equals K.cpp
//https://leetcode.com/problems/subarray-sum-equals-k/

/*
Given an array of integers nums and an integer k, return the total number of subarrays whose sum equals to k.
A subarray is a contiguous non-empty sequence of elements within an array.

Example 1:

Input: nums = [1,1,1], k = 2
Output: 2
*/


class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        unordered_map<int,int>mp;
        int ans=0,sum=0;
        mp[sum]++;
        for(int i=0;i<nums.size();i++){
            sum+=nums[i];
            ans+=mp[sum-k];
            mp[sum]++;
        }
        return ans;
    }
};

File: python/Prefix Sum/Prefix Sum/560. Subarray Sum Equals K.py
# https://leetcode.com/problems/subarray-sum-equals-k/description/

'''
Given an array of integers nums and an integer k, return the total number of subarrays whose sum equals to k.
A subarray is a contiguous non-empty sequence of elements within an array.

Example 1:
Input: nums = [1,1,1], k = 2
Output: 2
Example 2:
Input: nums = [1,2,3], k = 3
Output: 2

Constraints:
1 <= nums.length <= 2 * 104
-1000 <= nums[i] <= 1000
-107 <= k <= 107
'''

class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        freq=defaultdict(int)
        sum1=ans=0
        freq[sum1]+=1
        for i in range(len(nums)):
            sum1+=nums[i]
            ans+=freq[sum1-k]
            freq[sum1]+=1
        return ans
        

# Complexity Analysis
# Time complexity : O(n). The entire nums array is traversed only once.
# Space complexity : O(n). Hashmap map can contain up to n distinct entries in the worst case
